code/mission/missioncampaign.cpp

```diff
@@ -1079,10 +1083,59 @@ void mission_campaign_store_variables(int persistence_type, bool store_red_alert
 	}
 }
 
+// jg18 - adapted from mission_campaign_store_variables()
+void mission_campaign_store_containers(ContainerType persistence_type, bool store_red_alert)
+{
+	if (!sexp_container_has_persistent_non_eternal_containers()) {
+		// nothing to do
+		return;
+	}
+
+	if (store_red_alert) {
+		Campaign.persistent_containers = Campaign.red_alert_containers;
+	}
+
+	for (const auto &container : get_all_sexp_containers()) {
+		if (!container.is_eternal()) {
+			if (any(container.type & persistence_type)) {
+				// see if we already have a container with this name
+				auto cpc_it = std::find_if(Campaign.persistent_containers.begin(),
+					Campaign.persistent_containers.end(),
+					[container](const sexp_container& cpc) {
+						return !stricmp(container.container_name.c_str(), cpc.container_name.c_str());
+					});
+
+				if (cpc_it != Campaign.persistent_containers.end()) {
+					*cpc_it = container;
+				} else {
+					// new container
+					Campaign.persistent_containers.emplace_back(container);
+				}
+			}
+		} else if (any(persistence_type & ContainerType::SAVE_ON_MISSION_PROGRESS) &&
+				   any(container.type & persistence_type) && container.is_eternal()) {
+			// we might need to save some eternal player-persistent containers
+			auto ppc_it = std::find_if(Player->containers.begin(),
+				Player->containers.end(),
+				[container](const sexp_container& ppc) {
+					return !stricmp(container.container_name.c_str(), ppc.container_name.c_str());
```

Comment:
Same question as above, about whether container comparison here should include container type (see `sexp_container::type_matches()`) and not just name.